modules/database/database.bal

```diff
@@ -1019,40 +1031,45 @@ public isolated transactional function updateTimelineSubsteps(int stepId, Timeli
         if tmp != () {
-            oldDBSubstep = tmp.value;
+            oldSubstepNr += 1;
+            TimelineSubstepSQL oldDBSubstep = tmp.value;
             // make sure that no invalid changes have been made
             if receivedSubstep.href != oldDBSubstep.href {
-                return error(string `UI illegally changed the href of a substep or changed the order of substeps (stepId: ${stepId}, new hrefUi: ${receivedSubstep.href}).`);
+                return error(string `UI illegally changed the href of a substep or changed the order of substeps (stepId: ${stepId}, substep index: ${oldSubstepNr}, new hrefUi: ${receivedSubstep.href}).`);
             }
             if receivedSubstep.substepId != () && receivedSubstep.substepId != oldDBSubstep.substepId {
                 string? substepId = receivedSubstep.substepId;
-                return error(string `UI illegally set or changed the substepId of a substep  or changed the order of substeps (stepId: ${stepId}, new substepId: ${substepId != () ? substepId : ""}).`);
+                return error(string `UI illegally set or changed the substepId of a substep  or changed the order of substeps (stepId: ${stepId}, substep index: ${oldSubstepNr}, new substepId: ${substepId != () ? substepId : ""}).`);
             }
             if receivedSubstep.hrefUi != () && receivedSubstep.hrefUi != oldDBSubstep.hrefUi {
                 string? hrefUi = receivedSubstep.hrefUi;
-                return error(string `UI illegally set or changed the hrefUi of a substep or changed the order of substeps (stepId: ${stepId}, new hrefUi: ${hrefUi != () ? hrefUi : ""}).`);
+                return error(string `UI illegally set or changed the hrefUi of a substep or changed the order of substeps (stepId: ${stepId}, substep index: ${oldSubstepNr}, new hrefUi: ${hrefUi != () ? hrefUi : ""}).`);
             }
             if receivedSubstep.cleared == 0 && oldDBSubstep.cleared == 1 {
-                return error(string `Previously cleared substep was set to cleared=false or changed the order of substeps (stepId: ${stepId}, href: ${receivedSubstep.href})!`);
+                return error(string `Previously cleared substep was set to cleared=false or changed the order of substeps (stepId: ${stepId}, substep index: ${oldSubstepNr}, href: ${receivedSubstep.href})!`);
             }
         } else {
             // new substep
             check createTimelineSubstep(stepId, receivedSubstep);
+            newSubstepNr += 1;
             changes = true;
         }
     }
-    // set all previous substeps to cleared by default (should have been done, but we don't care here)
-    check clearTimelineSubsteps(stepId, receivedSubsteps.length() - 1);
+    // set all previous substeps except the latest one to cleared by default (should have been done, but we don't care here). The latest one should only be set manually. 
+    boolean someCleared = check clearTimelineSubsteps(stepId, oldSubstepNr + newSubstepNr - 1);
```

Comment:
`receivedSubsteps.length() - 1` (or `-2`) is ok here and more readable. But this edit does not capture the case when every received substep is marked as cleared. There must be a test to see if the last substep is also cleared and if so mark it as cleared in the DB.